### PR TITLE
Spotify artist import

### PR DESCRIPTION
**Pull request by hedgehog1833**
It is now possible to import artists from Spotify. The import takes all albums the user has liked on Spotify and then fetches the artists. The import button is currently placed on the MyArtists page, this is open for debate. One thing I'd like to have in the future is a mechanism that shows a list of artists to import in the frontend to check/uncheck and takes only those that you want. I'm also thinking of other ways to import, not by album but by really followed artists. But as I use Spotify more the way that I save albums rather than follow artists, that was may first approach here.
Regarding the "workflow": it is currently possible to click on import before authorizing, which obviously does not work. I am not really sure what the best way to prevent that would be. Maybe we don't need an authorization button at all and this is done automatically when clicking on import 🤷‍♂️ But this PR is big enough already, so one step after the other :)

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[Spotify Artist Import](https://trello.com/c/jTFeZAZc/220-spotify-artist-import)

**Comment by DanielW1987** [comment]
@hedgehog1833 
You have conflits here.

**Comment by DanielW1987** [comment]
Authorization and import should be done together and not separately. I would have liked to see the synchronization with Spotify in the profile. The reason for this is that the synchronization becomes more complex.

But these are all topics for later PRs.

File: webapp/src/main/java/rocks/metaldetector/service/artist/ArtistServiceImpl.java [hunk 1]
```diff
@@ -37,4 +39,21 @@ public boolean existsArtistByExternalId(String externalId, ArtistSource source)
...
+  }
+
+  @Override
+  public List<String> findNewArtistIds(List<String> artistIds) {
+    List<String> existingArtistIds = artistRepository.findAllByExternalIdIn(artistIds).stream()
+        .map(ArtistEntity::getExternalId)
+        .collect(Collectors.toList());
+    return artistIds.stream()
+        .filter(id -> !existingArtistIds.contains(id))
+        .collect(Collectors.toList());
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/artist/ArtistServiceImpl.java:57]:
```suggestion
    artistRepository.findAllByExternalIdIn(artistIds).stream()
        .map(ArtistEntity::getExternalId)
        .forEach(artistIds::remove);

    return artistIds;
```

File: spotify/src/main/java/rocks/metaldetector/spotify/facade/SpotifyServiceImpl.java [hunk 2]
```diff
@@ -27,45 +35,74 @@
...
   @Override
   public SpotifyUserAuthorizationDto getAccessToken(String code) {
-    SpotifyUserAuthorizationResponse response = spotifyAuthorizationClient.getUserAuthorizationToken(code);
-    return spotifyUserAuthorizationTransformer.transform(response);
+    SpotifyUserAuthorizationResponse response = authorizationClient.getUserAuthorizationToken(code);
+    return userAuthorizationTransformer.transform(response);
+  }
+
+  @Override
+  public List<SpotifyAlbumDto> importAlbums(String token) {
```
Comment (DanielW1987) [inline spotify/src/main/java/rocks/metaldetector/spotify/facade/SpotifyServiceImpl.java:86]: That's not really an import at this point. It's more a fetch of liked albums or somethin like that.

File: spotify/src/main/java/rocks/metaldetector/spotify/client/SpotifyImportClient.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package rocks.metaldetector.spotify.client;
+
+import rocks.metaldetector.spotify.api.imports.SpotfiyAlbumImportResult;
+
+public interface SpotifyImportClient {
+
+  SpotfiyAlbumImportResult importAlbums(String token, int offset);
+}
```
Comment (DanielW1987) [inline spotify/src/main/java/rocks/metaldetector/spotify/client/SpotifyImportClient.java:8]: The import term is misleading here. The import is made at a later step.

File: webapp/src/main/java/rocks/metaldetector/service/spotify/SpotifyArtistImportServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        .collect(Collectors.toList());
+  }
+
+  private void persistNewArtists(List<String> newArtistsIds) {
+    List<SpotifyArtistDto> spotifyArtistDtos = new ArrayList<>();
+    int totalPages = (int) Math.ceil((double) newArtistsIds.size() / (double) PAGE_SIZE);
+    for (int i = 1; i <= totalPages; i++) {
+      List<String> idsPerPage = slicingService.slice(newArtistsIds, i, PAGE_SIZE);
+      spotifyArtistDtos.addAll(spotifyService.searchArtistsByIds(idsPerPage));
+    }
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/spotify/SpotifyArtistImportServiceImpl.java:65]: The for loop should be done within `SpotifyServiceImpl`, because there is already a similar looping.

File: webapp/src/main/java/rocks/metaldetector/service/artist/FollowArtistServiceImpl.java [hunk 5]
```diff
@@ -41,13 +43,32 @@
...
+    );
+
+    if (!followActionRepository.existsByUserIdAndArtistId(currentUser.getId(), artistEntity.getId())) {
+      FollowActionEntity followAction = FollowActionEntity.builder()
+          .user(currentUser())
+          .artist(artistEntity)
+          .build();
+
+      followActionRepository.save(followAction);
+    }
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/artist/FollowArtistServiceImpl.java:69]: I think line 57 to 69 can be merged wirh the method above.

File: webapp/src/main/resources/static/ts/src/service/my-artists-render-service.ts [hunk 6]
```diff
@@ -90,16 +99,26 @@ export class MyArtistsRenderService extends AbstractRenderService<MyArtistsRespo
...
         })
     }
 
-    private attachPagination(paginationData: Pagination) {
+    private attachPagination(paginationData: Pagination): void {
         const paginationList = this.paginationComponent.render(paginationData);
         this.hostElement.insertAdjacentElement("beforeend", paginationList);
     }
+
+    private renderImportButton(): void {
```
Comment (DanielW1987) [inline webapp/src/main/resources/static/ts/src/service/my-artists-render-service.ts:115]: I think the rendering is already done?! It's more an add of event listener and the method should be called within the constructor.

**Review by DanielW1987 — COMMENTED** [review]
My comments for now. I have not looked at it in the frontend yet. But I will definitely do that on the weekend.

File: webapp/src/main/java/rocks/metaldetector/service/artist/ArtistServiceImpl.java (hunk 1 above)
Comment (NilsDammenhayn) [reply webapp/src/main/java/rocks/metaldetector/service/artist/ArtistServiceImpl.java:57]: Yeah I thought about that at first. But then the Stream has a side effect on a list that is unrelated to this stream. That feels kind of weird. If the list that is given into this method is changed we could also make this method void and have the same effect in the calling class. Also not nice... To get rid of the second stream we need a `disjunct` function, guava has something like that. But just for this case I don't want to pull the dependency...
Long story short: I'd like to leave it that way :)

File: webapp/src/main/java/rocks/metaldetector/service/artist/FollowArtistServiceImpl.java (hunk 5 above)
Comment (NilsDammenhayn) [reply webapp/src/main/java/rocks/metaldetector/service/artist/FollowArtistServiceImpl.java:69]: I just realized that I am not even using the new `follow()` method anymore, so it can go completely (and some other new stuff too...). I guess I wrote that in the beginning and didn't realize when it became obsolete. And I didn't test it... Very weird 🤷‍♂️

File: webapp/src/main/resources/static/ts/src/service/my-artists-render-service.ts (hunk 6 above)
Comment (DanielW1987) [reply webapp/src/main/resources/static/ts/src/service/my-artists-render-service.ts:115]: You forgot to address this issue.

**Review by DanielW1987 — APPROVED** [review]
